### PR TITLE
Implement parent channel fetching when checking thread permissions

### PR DESCRIPTION
**Pull request by rtm516**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [x] Internal code
- [ ] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: #1921

## Description

When trying to get the IPermissionContainer for a thread it now gets it from the parent to stop a ClassCastException.

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 1]
```diff
@@ -919,7 +919,7 @@ private void checkUser()
             if (isFromType(ChannelType.PRIVATE))
                 throw new IllegalStateException("Cannot delete another User's messages in a PrivateChannel.");
 
-            IPermissionContainer permChan = (IPermissionContainer) channel;
+            IPermissionContainer permChan = getPermissionContainer();
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:922]: I think you can just directly use `getGuildChannel()` instead, or no?

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 2]
```diff
@@ -1087,6 +1087,24 @@ public void setMentions(List<User> users, List<Member> members)
         return collection;
     }
 
+    private IPermissionContainer getPermissionContainer()
+    {
+        IPermissionContainer permChan;
+
+        // Check if the channel is a thread
+        if (isFromType(ChannelType.GUILD_NEWS_THREAD) || isFromType(ChannelType.GUILD_PUBLIC_THREAD) || isFromType(ChannelType.GUILD_PRIVATE_THREAD))
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java]: You should use `ChannelType#isThread` for these checks usually, but using guild channel makes this helper method unnecessary

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java (hunk 1 above)
Comment (rtm516) [reply src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:922]: No, ends up with the same error.

Comment (MinnDevelopment) [reply src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:922]: I do not get an error trying to do this:

```kt
val channel = event.channel as GuildChannel // in a thread
val bool = event.member!!.hasPermission(channel, Permission.MESSAGE_SEND)
println(bool) // true
```

Comment (DV8FromTheWorld) [reply src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:922]: `hasPermission` accepts `GuildChannel` and knows to call `GuildChannel#getPermissionContainer()`

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 3]
```diff
@@ -940,7 +940,7 @@ else if (!getGuild().getSelfMember().hasPermission(permChan, Permission.MESSAGE_
             if (isFromType(ChannelType.PRIVATE))
                 throw new PermissionException("Cannot suppress embeds of others in a PrivateChannel.");
 
-            IPermissionContainer permChan = (IPermissionContainer) channel;
+            IPermissionContainer permChan = getPermissionContainer();
             if (!getGuild().getSelfMember().hasPermission(permChan, Permission.MESSAGE_MANAGE))
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:944]: Same statement as above. We don't need a concrete instance of `IPermissionContainer` here.

File: src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java [hunk 4]
```diff
@@ -1087,6 +1087,24 @@ public void setMentions(List<User> users, List<Member> members)
...
+            // Get the parent channel for permissions if we are using a thread
+            permChan = ((ThreadChannel) channel).getParentChannel();
+        }
+        else
+        {
+            permChan = (IPermissionContainer) channel;
+        }
+
+        return permChan;
+    }
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/internal/entities/ReceivedMessage.java:1106]: This is entirely unneeded. `GuildChannel#getPermissionContainer() : IPermissionContainer` exists and should be used when needing to get an `IPermissionContainer` in unknown concrete channel situations. 

However, we don't actually need an `IPermissionContainer` as stated above.

**Review by DV8FromTheWorld — CHANGES_REQUESTED** [review]

**Review by MinnDevelopment — APPROVED** [review]

**Review by DV8FromTheWorld — APPROVED** [review]

**Pull request merged**